ports/rp2/machine_uart.c

```diff
@@ -422,6 +448,9 @@ static mp_obj_t mp_machine_uart_make_new(const mp_obj_type_t *type, size_t n_arg
 
     // Get static peripheral object.
     machine_uart_obj_t *self = (machine_uart_obj_t *)&machine_uart_obj[uart_id];
+    self->mp_irq_obj = NULL;
+    MP_STATE_PORT(rp2_uart_irq_obj[self->uart_id]) = self->mp_irq_obj;
```

Comment:
Does this irq obj need to be cleared?  Usually something like `machine.UART(1)` is used to retrieve an existing object and not reset it.  But here, the IRQ will be reset (actually the hardware IRQ stays active, if it's enabled).

Reply:
The behavior of `machine.UART(1)` is not consistent over the  ports. Ig a port has static UART objects like RP2 or NRF, it will refer to the existing state. If a port has dynamic objects like STM32, ESP32, MIMXRT, Renesas, SAMD,  will instantiate a new object. It seems to be more consistent to see it as a new object and clear previous states.

Reply:
> If a port has static UART objects like RP2 or NRF, it will refer to the existing state.

Yes, exactly, on rp2 it will refer to the existing instance.  And so in this case (eg accessing `machine.UART(1)`) the IRQ settings should not be changed, for consistency.

> If a port has dynamic objects like STM32, ESP32, MIMXRT, Renesas, SAMD, will instantiate a new object.

Actually, on stm32 any previous UART instance is also retrieved unchanged with a `machine.UART(1)`.

> Yes, it needs to be cleared to have a clean state after soft reset

Then that probably needs to be done in a `machine_uart_deinit()` function.

Reply:
It can indeed be done there. Thanks for the hint. But uart.deinit() is not automatically called on soft reset.

Reply:
I could call `machine_uart_deinit()` in `mp_machine_uart_make_new` to settle the ground.

Reply:
I would think the UART can behave the same way as Pin, in the way IRQs are handled, and the way they are deinitialised on soft reset.

Reply:
So I added a `machine_uart_deinit_all()`  function and call it in main() upon soft reset,
and cleared the uart.irq() state in uart.deinit().

Reply:
Great, this is much better, thanks!